src/Status.cxx

```diff
@@ -0,0 +1,154 @@
+#include <cmath>
+
+#include <rapidjson/document.h>
+
+#include "Status.hpp"
+
+double average(const double &sum, const double &N) { return sum / N; }
+
+double standard_deviation(const double &sum, const double &sum_squared,
+                          const double &N) {
+  auto variance = (N * sum_squared - sum * sum) / (N * (N - 1));
+  if (variance > 0) { // due to numerical instabilities
+    return std::sqrt(variance);
+  } else {
+    return 0.0;
+  }
+}
+
+const std::pair<double, double>
+FileWriter::Status::message_size(const FileWriter::Status::MessageInfo &value) {
+  if (value.messages().first == 0) { // nan causes failure in JSON
+    return std::pair<double, double>{};
+  }
+  std::pair<double, double> result{
+      average(value.Mbytes().first, value.messages().first),
+      standard_deviation(value.Mbytes().first, value.Mbytes().second,
+                         value.messages().first)};
+  return result;
+}
+
+const std::pair<double, double> FileWriter::Status::message_frequency(
+    const FileWriter::Status::MessageInfo &value,
+    const double time_difference) {
+  if (time_difference < 1e-10) {
+    return std::pair<double, double>({0, 0});
+  }
+  std::pair<double, double> result{
+      average(value.messages().first, time_difference),
+      standard_deviation(value.messages().first, value.messages().second,
+                         time_difference)};
+  return result;
+}
+
+const std::pair<double, double> FileWriter::Status::message_throughput(
+    const FileWriter::Status::MessageInfo &value,
+    const double time_difference) {
+  if (time_difference < 1e-10) {
+    return std::pair<double, double>({0, 0});
+  }
+  std::pair<double, double> result{
+      average(value.Mbytes().first, time_difference),
+      standard_deviation(value.Mbytes().first, value.Mbytes().second,
+                         time_difference)};
+  return result;
+}
+
+//  \author Michele Brambilla <michele.brambilla@psi.ch>
+//  \date Wed Oct 11 14:47:17 2017
+
+// data-race safe += with double
+template <class T> void atomic_add(std::atomic<T> &value, const T &other) {
+  auto current = value.load();
+  while (!value.compare_exchange_weak(current, current + other))
+    ;
```

Comment:
Isn't this a potential dead-lock? because `value` is never reloaded on failed CAS.

`atomic_fetch_add` would probably suffice?

Reply:
Unfortunately ```atomic_fetch_add``` wants integers. Indeed you're right, current has to be reloaded upon failure.